feat: new compilation filter for Power-assert (#5729)
JetBrains/kotlin-web-site: docs/topics/compiler-plugins/power-assert.md

The Power-assert plugin provides several options to customize its behavior:

> * **`functions`** lists fully qualified paths to functions whose calls the Power-assert plugin transforms. If not specified, only `kotlin.assert()` calls are transformed.

Reviewer: ```suggestion
* **`functions`** lists the fully qualified paths to functions that the Power-assert plugin transforms when called. If not specified, the plugin transforms only `kotlin.assert()` calls.
```
I find "whose calls" a bit hard to read, though I can't put a finger on exactly why. It's probably the nesting. I suggested an alternative that avoids that. I also suggested making the second sentence active.

Author: I like it, but I think 'when called' may sound like 'when the plugin is called', so I changed it a bit

The Power-assert plugin provides several options to customize its behavior:

* **`functions`** lists fully qualified paths to functions whose calls the Power-assert plugin transforms. If not specified, only `kotlin.assert()` calls are transformed.
* **`includedSourceSets`** lists Gradle source sets that the Power-assert plugin transforms. If not specified, all _test source sets_ are transformed.

> * **`compilationFilter`** controls if the Power-assert plugin is applied to all Kotlin compilations or only to test compilations. If not specified, the plugin is applied to all compilations.

Reviewer 2: These 2 options are mutually exclusive. If `includedSourceSets` is specified, the `compilationFilter` is ignored. `includedSourceSets` is now deprecated, so it should not be recommended. You can achieve the same behavior with `compilationFilter` like this:
```kotlin
powerAssert {
  compilationFilter = PowerAssertCompilationFilter {
    it.name in setOf("commonMain", "jvmMain", "jsMain", "nativeMain")
  }
}
```

Author: Thank you for the clarification! added a note about the deprecation

> * **`compilationFilter`** controls if the Power-assert plugin is applied to all Kotlin compilations or only to test compilations. If not specified, the plugin is applied to all compilations.

Reviewer 2: I would say that it controls to which Kotlin compilations Power-assert plugin is applied. It's not limited to all or tests, there's a lot more control since you can create your own custom filter.

And the default for `compilationFilter` is `PowerAssertCompilationFilter.TESTS`, so all _test source sets_, just like it is for `includedSourceSets`.
